### ISSUE TITLE
Pylance reports "Type parameter is invariant" for List[TypedDict], requires explicit annotation for variable assignment—can this warning be suppressed or improved?

### ISSUE THREAD

**Issue by wangzhe258369** [microsoft/pyright#10908]
Hi,  

When using type annotations like `List[MyPayload]` where `MyPayload` is a `TypedDict`, Pylance reports the following warning and requires explicit type annotation for variable assignment:

```
Type parameter "_T@list" is invariant, but "list[dict[str, Any]]" is not assignable to "List[MyPayload]"
  Consider switching from "list" to "Sequence" (covariant)
```

**Example:**
```python
from typing import List, TypedDict

class MyPayload(TypedDict):
    id: int
    name: str

def foo(x: List[MyPayload]) -> None:
    pass

my_payload = [
    {"id": 1, "name": "a"},
    {"id": 2, "name": "b"},
]

foo(my_payload)  # Pylance reports: List[Dict[str, Any]] is not assignable to List[MyPayload]
```

However, the design intent of TypedDict is to allow dicts that match the specified structure to be used as instances of the TypedDict. It seems reasonable that as long as each dict in `my_payload` matches `MyPayload`, the assignment should be valid.

Currently, to avoid the warning, I have to write:
```python
my_payload: List[MyPayload] = [
    {"id": 1, "name": "a"},
    {"id": 2, "name": "b"},
]

foo(my_payload)
```
But adding explicit type annotations for every such variable assignment is cumbersome and reduces code readability.

**Questions:**
1. Is there a way to suppress or disable this warning in Pylance?
2. Can Pylance be improved to recognize that a list of dicts matching the TypedDict structure is assignable to `List[TypedDict]` without explicit annotation?
3. Is there a recommended way to annotate such variables more conveniently?

**Expected:**  
No warning, or at least an option to suppress this warning when using TypedDict in lists, as long as the dicts match the structure.

**Actual:**  
Pylance reports the invariance warning and requires explicit annotation, which is inconvenient for common usage patterns.

Thank you!

**Comment by erictraut** [comment]
Pyright is working as intended here, so this isn't a bug.

When you use a dictionary literal expression in your code, by default pyright infers a `dict` type. This is what most users assume, and it results in correct type checking behaviors most of the time. For more details about dict expression inference, refer to [this documentation](https://microsoft.github.io/pyright/#/type-inference?id=dictionary-expressions).

Default inference rules produce the correct type inference most of the time, but there will always be situations where you want to override these default behaviors. To do this, you can use [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types). This involves explicit type annotations, as you've noted above.

The type of a dict or list must be established at the time it is constructed. All subsequent operations on that object (or aliases to that object) must then honor that type. So it wouldn't be safe for a type checker to initially infer the type as `list[dict]` and then later treat it as `list[MyPayload]`. The object needs to "start its life" as a `list[MyPayload]`.

There are a couple of small things you can do to reduce the burden of adding the explicit type annotation:
1. Use `list` rather than `List` in the type annotation. The latter is no longer needed as of Python 3.10. This eliminates the need to import an extra symbol.
2. If you find that you're using `list[MyPayload]` repeatedly, define a type alias to reduce the verbosity.
